### DISCUSSION TITLE
Dependencies doesn't respect order

### DISCUSSION THREAD

**Discussion by andriytam** [fastapi/fastapi#9932] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, Depends, Header, APIRouter
from pydantic import BaseModel


class Item(BaseModel):
    id: str
    name: int


router = APIRouter()


def item_already_exists(item: Item):
    print("Checking item in DB.")
    return item


@router.post("/item", dependencies=[Depends(item_already_exists)])
def create_item(item: Item) -> Item:
    return item


def require_header(must_have_header: Annotated[str, Header()]):
    print("Checking header in REQUEST.")


app = FastAPI()
app.include_router(router, prefix="/v1", dependencies=[Depends(require_header)])
```

### Description

I do have a case where I want to validate header and than validate if item exists in DB. If header is not provided I expect FastAPI not to call nested dependencies.

Request with `must-have-header` works correctly, but if it's not provided the dependencies of nested router are run before dependencies of the APP.

Here is an example:

Making request without `must-have-header`:
```
curl -X 'POST'   'http://127.0.0.1:8000/v1/item'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "id": "string",
  "name": 0
}'
```

Response:
```
{"detail":[{"type":"missing","loc":["header","must-have-header"],"msg":"Field required","input":null,"url":"https://errors.pydantic.dev/2.0.3/v/missing"}]}
```

Logs:
```
INFO:     127.0.0.1:55740 - "POST /v1/item HTTP/1.1" 200 OK
Checking item in DB.
INFO:     127.0.0.1:38980 - "POST /v1/item HTTP/1.1" 422 Unprocessable Entity
```

### Operating System

Linux

### FastAPI Version

0.100.0

### Python Version

Python 3.10.8

**Comment by jgould22** [comment, marked as answer]
It may seem intuitive that a router level dependency would be a dependency for any route level request but I am not sure that is how it works. 

You have created a dependency tree like this 
```mermaid
graph TD
    item_already_exists --> create_item
    require_header --> create_item
```

when what you probably wanted was 
```mermaid
graph TD
    require_header --> item_already_exists --> create_item
```

The immediate solution would be to tell fastapi that `require_header` is required for  `item_already_exists` by adding `require_header` as a dependency.

**Comment by andriytam** [comment]
Thanks for quick response.

However I still thinks order is not respected or at least not specified expected to be that way.

In case of:
 ```mermaid
 graph TD
     item_already_exists --> create_item
     require_header --> create_item
 ```
if you provide curl request header: `-H 'must-have-header: value'` the order is:
```
require_header
item_already_exists
```
if not provided than order is:
```
item_already_exists
require_header
```




The work around that I found it to make `header` optional:
```
def require_header(must_have_header: Annotated[str, Header()] = None):
```
however it will act as required and probably needs some tweaks with openapi schema to specify that header as required.

**Reply by jgould22** [reply]
> However I still thinks order is not respected or at least not specified expected to be that way. 

There is no order in that example and any observed order is an implementation detail and can change as the implementation changes.

If you must execute them in a specific order you should make one a dependency of the other.

**Reply by andriytam** [reply]
In your example it's OK for single endpoint. I do have many of them and wrapping each one of them doesn't looks handy and become hard to manage and keep well tested.

How to wrap entire `router` into?

 ```mermaid
graph TD
    require_header --> item_already_exists --> create_item
```
